design/object-store.md

```diff
@@ -0,0 +1,141 @@
+1. The data pool is created (`my-store.rgw.buckets.data`)
+1. A Ceph realm is created 
+1. A Ceph zone group is created in the new realm
+1. A Ceph zone is created in the new zone group
+1. A cephx key is created for the rgw daemon
+1. A Kubernetes service is created to provide load balancing for the RGW pod(s)
+1. A Kubernetes deployment is created to start the RGW pod(s) with the settings for the new zone
+
+When the RGW pods start, the object store is ready to receive the http or https requests as configured.
+
+
+## Object Store CRD
+
+The object store settings are exposed to Rook as a Custom Resource Definition (CRD). The CRD is the Kubernetes-native means by which the Rook operator can watch for new resources. The operator stays in a control loop to watch for a new object store, changes to an existing object store, or requests to delete an object store. 
+
+### Pools
+
+The pools are the backing data store for the object store and are created with specific names to be private to an object store. Pools can be configured with all of the settings that can be specified in the [Pool CRD](/Documentation/pool-crd.md). The underlying schema for pools defined by a pool CRD is the same as the schema under the `metadataPool` and `dataPool` elements of the object store CRD. All metadata pools are created with the same settings, while the data pool can be created with independent settings. The metadata pools must use replication, while the data pool can use replication or erasure coding. 
+
+```yaml
+  metadataPool:
+    failureDomain: host
+    replicated:
+      size: 3
+  dataPool:
+    failureDomain: device
+    erasureCoded:
+      dataChunks: 6
+      codingChunks: 2
+```
+
+### Gateway
+
+The gateway settings correspond to the RGW service. 
+- `type`: Can be `s3`. In the future support for `swift` can be added.
+- `sslCertificateRef`: If the certificate is not specified, SSL will not be configured. If specified, this is the name of the Kubernetes secret that contains the SSL certificate to be used for secure connections to the object store. Rook will look in the secret provided at the `cert` key name. The value of the `cert` key must be in the format expected by the [RGW service](http://docs.ceph.com/docs/master/install/install-ceph-gateway/#using-ssl-with-civetweb): "The server key, server certificate, and any other CA or intermediate certificates be supplied in one file. Each of these items must be in pem form."
+- `port`: Where the RGW pods and the RGW service will be listening
+- `replicas`: The number of pods that will be started for this object store
+- `placement`: The rgw pods can be given standard Kubernetes placement restrictions with `nodeAffinity`, `tolerations`, `podAffinity`, and `podAntiAffinity` similar to placement defined for daemons configured by the [cluster CRD](/cluster/examples/kubernetes/rook-cluster.yaml).
```

Comment:
really feels like we are creating an abstraction here ontop of something the kubernetes does well -- replication controllers. should we just let the users create a standard deployment with a replication controller, etc. for rgw?

Reply:
Are you suggesting that creating an object store would be two steps?
1. Create the object store CRD
2. Create the rgw deployment spec
That's a good question since you may want full control of your rgw deployment. We could ask the same question for mons, osds, and other daemons. Where do we draw the line between making this simple and automating it with the operator, compared to giving full flexibility for the more advanced users?

It feels like if we split this into multiple steps, the end user experience would be fragmented and difficult to manage, instead of the operator giving us a full e2e solution.

Reply:
the flip side is that you have to expose in this CRD some of the concepts that already exist in K8S.

Reply:
also compared to mon and odds the rgw pods are stateless. if we said we don't care about a single CRD experience, but instead a single yaml file with multiple resources, what would this look like? Worth modeling IMO.

Reply:
alternatively we would need to expose most of the features of replication controllers, auto-scalers, etc. this is a tough one.

Reply:
The challenge with the single yaml with multiple resources is that you don't want to start the rgw service until all the pools, the cephx key, realm, zonegroup, and zone are created. We would have to implement logic in the rgw pod to wait for those resources to exist. Today the daemons don't make any kubernetes calls or even know they're running in kubernetes for that matter.